src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+                }
+            }
+
+            public TOperation Tail
+            {
+                get
+                {
+                    Debug.Assert(!IsStopped);
+                    return (TOperation)_tail;
+                }
+            }
+
+            public void Enqueue(TOperation operation)
+            {
+                Debug.Assert(!IsStopped);
+
+                if (!IsEmpty)
+                {
+                    operation.Next = _tail.Next;
+                    _tail.Next = operation;
+                }
+
+                _tail = operation;
+            }
+
+            public void Dequeue()
+            {
+                Debug.Assert(!IsStopped);
+                Debug.Assert(!IsEmpty);
+
+                AsyncOperation head = _tail.Next;
+                if (head == _tail)
+                {
+                    _tail = null;
+                }
+                else
+                {
+                    _tail.Next = head.Next;
+                }
+            }
+
+            public Queue<TOperation> Stop()
+            {
+                Queue<TOperation> result = this;
+                _tail = null;
+                State = State.Stopped;
+                return result;
+            }
+        }
+
+        int _fileDescriptor;
+        GCHandle _handle;
+        Queue<TransferOperation> _receiveQueue;
+        Queue<TransferOperation> _sendQueue;
+        Queue<AcceptOrConnectOperation> _acceptOrConnectQueue;
+        object _closeLock = new object();
+        object _queueLock = new object();
+        SocketAsyncEngine _engine;
+        uint _registeredEvents;
+
+        public SocketAsyncContext(int fileDescriptor, SocketAsyncEngine engine)
+        {
+            _fileDescriptor = fileDescriptor;
+            _engine = engine;
+
+            var events = Interop.libc.EPOLLIN | Interop.libc.EPOLLOUT | Interop.libc.EPOLLRDHUP | Interop.libc.EPOLLET;
+
+            Interop.Error errorCode;
+            if (!_engine.TryRegister(this, _fileDescriptor, events, 0, ref _handle, out errorCode))
+            {
+                throw new Exception(string.Format("SocketAsyncContext: {0}", errorCode));
+            }
+
+            _registeredEvents = events;
+        }
+
+        public void Close()
+        {
+            lock (_closeLock)
+            lock (_queueLock)
+            {
+                // Force a HUP event in order to drain the queues.
+                HandleEvents(Interop.libc.EPOLLHUP);
+            }
+        }
+
+        private bool TryBeginOperation<TOperation>(ref Queue<TOperation> queue, TOperation operation, out bool isStopped)
+            where TOperation : AsyncOperation
+        {
+            lock (_queueLock)
+            {
+                switch (queue.State)
+                {
+                    case State.Stopped:
+                        isStopped = true;
+                        return false;
+
+                    case State.Clear:
+                        break;
```

Comment:
You have this here just to be explicit about all of the cases?  If so, consider instead doing:
```C#
default:
    Debug.Assert(queue.State == State.Clear);
    break;
```